skynet.go

```diff
@@ -57,33 +132,259 @@ type (
 )
 
 const (
+	// DefaultPortalURL is the default URL of the portal to use.
+	DefaultPortalURL = "https://siasky.net"
+
 	// URISkynetPrefix is the URI prefix for Skynet.
 	URISkynetPrefix = "sia://"
 )
 
 var (
+	// DefaultConnectionOptions contains the default connection options.
+	DefaultConnectionOptions = ConnectionOptions{
+		PortalURL: DefaultPortalURL,
+	}
+
+	// DefaultAddSkykeyOptions contains the default addskykey options.
+	DefaultAddSkykeyOptions = AddSkykeyOptions{
+		ConnectionOptions:   DefaultConnectionOptions,
+		PortalAddSkykeyPath: "/skynet/addskykey",
+	}
+	// DefaultCreateSkykeyOptions contains the default createskykey options.
+	DefaultCreateSkykeyOptions = CreateSkykeyOptions{
+		ConnectionOptions:      DefaultConnectionOptions,
+		PortalCreateSkykeyPath: "/skynet/createskykey",
+	}
+	// DefaultGetSkykeyIDOptions contains the default skykeyid GET options.
+	DefaultGetSkykeyIDOptions = GetSkykeyIDOptions{
+		ConnectionOptions:     DefaultConnectionOptions,
+		PortalGetSkykeyIDPath: "/skynet/skykeyid",
+	}
+	// DefaultGetSkykeyOptions contains the default skykey GET options.
+	DefaultGetSkykeyOptions = GetSkykeyOptions{
+		ConnectionOptions:   DefaultConnectionOptions,
+		PortalGetSkykeyPath: "/skynet/skykey",
+	}
+	// DefaultListSkykeysOptions contains the default skykeys options.
+	DefaultListSkykeysOptions = ListSkykeysOptions{
+		ConnectionOptions:     DefaultConnectionOptions,
+		PortalListSkykeysPath: "/skynet/skykeys",
+	}
+
 	// DefaultDownloadOptions contains the default download options.
 	DefaultDownloadOptions = DownloadOptions{
-		PortalURL:          "https://siasky.net",
+		ConnectionOptions:  DefaultConnectionOptions,
 		PortalDownloadPath: "/",
 	}
 
 	// DefaultUploadOptions contains the default upload options.
 	DefaultUploadOptions = UploadOptions{
-		PortalURL:                    "https://siasky.net",
+		ConnectionOptions:            DefaultConnectionOptions,
 		PortalUploadPath:             "/skynet/skyfile",
 		PortalFileFieldName:          "file",
 		PortalDirectoryFileFieldName: "files[]",
 	}
+
+	// ErrResponseError is the error for a response with a status code >= 400.
+	ErrResponseError = errors.New("error response")
 )
 
+// AddSkykey stores the given base-64 encoded skykey with the renter's skykey
+// manager.
+func AddSkykey(skykey string, opts AddSkykeyOptions) error {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalAddSkykeyPath)
+	url = fmt.Sprintf("%s?skykey=%s", url, skykey)
+
+	req, err := makeRequest(opts.ConnectionOptions, "POST", url, body)
+	if err != nil {
+		return errors.AddContext(err, "could not make request")
+	}
+
+	// Add the skykey.
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.AddContext(err, "could not execute POST")
+	}
+	if resp.StatusCode >= 400 {
```

Comment:
Do we expected to receive non error codes other that successful? 

My initial reaction is that we should now the StatusCodes we should receive and handle them accordingly.  Something like a switch statement that handles different codes explicitly and gives a developer error if we have missed one?

Otherwise if that is over the top I think we should explicitly check for success and return an error if not so we aren't hiding any weird status codes that might create UX bugs or inconsistencies for users. 

Thoughts?

Reply:
Yeah, I agree. The potential errors are part of the public API, after all. I'll do this on Monday

Reply:
Hmm so I'm actually not sure what checking for explicit codes would give us. I'm not a web dev person but I'm pretty sure that all codes >= 400 are defined to be errors. Any other code should therefore be a success. I don't think we have many success codes at all but I hesitate to enumerate them here because I'm thinking that a portal could define its own (e.g. redirects which should have their own status code). Maybe @kwypchlo can weigh in here but I'm thinking we move this to a followup so as not to block the PR.

Reply:
a follow up is fine.

Reply:
As far as I know, we only use 200 (with response content) and 204 (without content) as OK but any 2xx should be treated as ok while any 4xx and 5xx should be treated as error. We don't have a finite list of what errors does the API return, that would have to be up to the api to document first.